aws/src/main/java/org/apache/iceberg/aws/s3/signer/S3V4RestSignerClient.java

```diff
@@ -0,0 +1,341 @@
+    // so we need to clear the current path from the request
+    SdkHttpFullRequest.Builder mutableRequest = request.toBuilder();
+    mutableRequest.encodedPath("");
+    mutableRequest.uri(signedComponent.signedURI());
+    reconstructHeaders(signedComponent.headers(), mutableRequest);
+
+    return mutableRequest.build();
+  }
+
+  private void reconstructHeaders(
+      Map<String, List<String>> signedAndUnsignedHeaders,
+      SdkHttpFullRequest.Builder mutableRequest) {
+    Map<String, List<String>> headers = Maps.newHashMap(signedAndUnsignedHeaders);
+    // we need to remove the Cache-Control header that is being sent by the server
+    headers.remove(CACHE_CONTROL);
+
+    // we need to overwrite whatever headers the server signed/unsigned with the ones from the
+    // original request and then put all headers back to the request
+    headers.putAll(mutableRequest.headers());
+    headers.forEach(mutableRequest::putHeader);
+  }
+
+  private boolean canBeCached(S3SignResponse s3SignResponse) {
+    List<String> cacheControlHeader = s3SignResponse.headers().get(CACHE_CONTROL);
+    if (null != cacheControlHeader) {
+      LOG.debug("Received {} header: {}", CACHE_CONTROL, cacheControlHeader);
+
+      boolean canBeCached = cacheControlHeader.contains(CACHE_CONTROL_PRIVATE);
+      if (!canBeCached && !cacheControlHeader.contains(CACHE_CONTROL_NO_CACHE)) {
+        LOG.warn("Unsupported {} header: {}", CACHE_CONTROL, cacheControlHeader);
+      }
+      return canBeCached;
+    }
+    return false;
+  }
+
+  private AuthSession newSession(String token, AuthSession parent) {
+    AuthSession session =
+        new AuthSession(parent.headers(), token, OAuth2Properties.ACCESS_TOKEN_TYPE);
+    AuthSession.scheduleTokenRefresh(
+        httpClient(),
+        TOKEN_REFRESH_EXECUTOR,
+        session,
+        System.currentTimeMillis(),
+        tokenExpiresInMs(),
+        TimeUnit.MILLISECONDS);
+    return session;
+  }
+
+  private AuthSession newSession(OAuthTokenResponse response, AuthSession parent) {
+    AuthSession session =
+        new AuthSession(parent.headers(), response.token(), response.issuedTokenType());
+    if (response.expiresInSeconds() != null) {
+      AuthSession.scheduleTokenRefresh(
+          httpClient(),
+          TOKEN_REFRESH_EXECUTOR,
+          session,
+          System.currentTimeMillis(),
+          response.expiresInSeconds(),
+          TimeUnit.SECONDS);
+    }
+    return session;
+  }
+
+  private AuthSession newSessionWithImmediateRefresh(String token) {
+    AuthSession session =
+        new AuthSession(ImmutableMap.of(), token, OAuth2Properties.ACCESS_TOKEN_TYPE);
+    if (null != token) {
+      AuthSession.scheduleTokenRefresh(
+          httpClient(),
+          TOKEN_REFRESH_EXECUTOR,
+          session,
+          System.currentTimeMillis(),
+          0,
+          TimeUnit.MILLISECONDS);
+    }
+
+    return session;
```

Comment:
If the token is expired (e.g. a task is restored in Flink), then it looks like this will create a session with an expired token, and also schedule a refresh of the expired token, which should fail.